common/chains/client/multi_node_client.go

```diff
@@ -0,0 +1,516 @@
+	SUB types.Subscription,
+] interface {
+	// Select returns a Node, or nil if none can be selected.
+	// Implementation must be thread-safe.
+	Select() Node[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
+	// Name returns the strategy name, e.g. "HighestHead" or "RoundRobin"
+	Name() string
+}
+
+type MultiNodeClient[
+	CHAIN_ID types.ID,
+	SEQ types.Sequence,
+	ADDR types.Hashable,
+	BLOCK_HASH types.Hashable,
+	TX any,
+	TX_HASH types.Hashable,
+	EVENT any,
+	EVENT_OPS any, // event filter query options
+	TX_RECEIPT types.ChainReceipt[TX_HASH, BLOCK_HASH],
+	FEE feetypes.Fee,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+] interface {
+	RPCClient[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
+	Dial(context.Context) error
+	Close() error
+	NodeStates() map[string]string
+	IsL2() bool
+	BatchCallContextAll(ctx context.Context, b []any) error
+	runLoop()
+	nLiveNodes() (int, int64, *utils.Big)
+	report()
+}
+
+func ContextWithDefaultTimeout() (ctx context.Context, cancel context.CancelFunc) {
+	return context.WithTimeout(context.Background(), queryTimeout)
+}
+
+type multiNodeClient[
+	CHAIN_ID types.ID,
+	SEQ types.Sequence,
+	ADDR types.Hashable,
+	BLOCK_HASH types.Hashable,
+	TX any,
+	TX_HASH types.Hashable,
+	EVENT any,
+	EVENT_OPS any, // event filter query options
+	TX_RECEIPT types.ChainReceipt[TX_HASH, BLOCK_HASH],
+	FEE feetypes.Fee,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+] struct {
+	utils.StartStopOnce
+	nodes               []Node[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
+	sendonlys           []Node[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
+	chainID             CHAIN_ID
+	chainType           config.ChainType
+	logger              logger.Logger
+	selectionMode       string
+	noNewHeadsThreshold time.Duration
+	nodeSelector        NodeSelector[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
+
+	activeMu   sync.RWMutex
+	activeNode Node[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
+
+	chStop utils.StopChan
+	wg     sync.WaitGroup
+}
+
+func NewMultiNodeClient[
+	CHAIN_ID types.ID,
+	SEQ types.Sequence,
+	ADDR types.Hashable,
+	BLOCK_HASH types.Hashable,
+	TX any,
+	TX_HASH types.Hashable,
+	EVENT any,
+	EVENT_OPS any, // event filter query options
+	TX_RECEIPT types.ChainReceipt[TX_HASH, BLOCK_HASH],
+	FEE feetypes.Fee,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+](logger logger.Logger, selectionMode string, noNewHeadsTreshold time.Duration, nodes []Node[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB], sendonlys []Node[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB], chainID CHAIN_ID, chainType config.ChainType,
```

Comment:
nit: break this line into multiple lines for better readability

Reply:
done